File: src/main/java/net/tassia/pancake/cli/PancakeCLI.java
package net.tassia.pancake.cli;

import net.tassia.pancake.Pancake;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class PancakeCLI {
	private final ExecutorService service;
	private final CLIReader reader;

	/* Constructor */
	public PancakeCLI(Pancake pancake) {
		this.reader = new CLIReader(pancake);
		this.service = Executors.newSingleThreadExecutor();

		reader.commands.put("sessions", new CMD_Sessions());
		reader.commands.put("end", COMMAND_END);
	}
	/* Constructor */



	/* Start & Stop */
	public void start() {
		service.submit(reader);
	}

	public void stop() {
		reader.scanner.close();
		service.shutdownNow();
	}
	/* Start & Stop */





	/* End Command */
	public static final CLICommand COMMAND_END = (pancake, args) -> {
		pancake.getLogger().info("Good bye!");
		System.exit(0);
	};
	/* End Command */

}

File: src/main/java/net/tassia/pancake/cli/PancakeCLI.kt
package net.tassia.pancake.cli

import net.tassia.pancake.Pancake

class PancakeCLI(pancake: Pancake) {

	private val commands: MutableMap<CommandInfo, Command> = mutableMapOf()
	private val listener: CommandListener = CommandListener(pancake) { search ->
		commands.entries.find {
			it.key.name.equals(search, true)
		}?.value
	}
	private val thread: Thread = Thread(listener, "PancakeCLI").also { it.start() }



	fun stop() {
		listener.listening = false
		if (thread.isAlive) {
			thread.interrupt()
		}
	}



	fun commands(): Set<CommandInfo> = commands.keys

	fun add(info: CommandInfo, command: Command) = commands.set(info, command)

	fun remove(info: CommandInfo) = commands.remove(info)

}
